### PR TITLE
feat: add json filtering functions/compares (pql, mysql)

### PR DESCRIPTION
**Pull request by Weakky**
## Overview

Adds the following functions/compares:

- (fun) `json_extract`: extract a subset of a JSON blob given a JSON path/array of keys
- (compare) `json_array_contains`
- (compare) `json_array_not_contains`
- (compare) `json_array_begins_with`
- (compare) `json_array_not_begins_with`
- (compare) `json_array_ends_into`
- (compare) `json_array_not_ends_into`
- (compare) `json_type_equals`

- Works only on MySQL/PostgreSQL for now.
- SQLite can be added in the future once it supports the JSON type. The migration team has work to do first, which needs to be planned
- MSSQL can also be added in the future once it supports the JSON type.

### CODE REVIEW

File: src/visitor/mysql.rs [hunk 1]
```diff
@@ -278,6 +278,133 @@ impl<'a> Visitor<'a> for Mysql<'a> {
...
+        left: Expression<'a>,
+        right: Expression<'a>,
+        not: bool,
+    ) -> visitor::Result {
+        self.write("JSON_EXTRACT(")?;
+        self.visit_expression(left)?;
+        self.write(", ")?;
+        self.visit_parameterized(Value::text("$[0]"))?;
+        self.write(")")?;
+        if not {
```
Comment (pimeys) [inline src/visitor/mysql.rs:334]: nitpick: spacing before and after this block

File: src/visitor/mysql.rs [hunk 2]
```diff
@@ -278,6 +278,133 @@ impl<'a> Visitor<'a> for Mysql<'a> {
...
+        self.write("CONCAT('$[', ")?;
+        self.write("JSON_LENGTH(")?;
+        self.visit_expression(left)?;
+        self.write(") - 1, ']'))")?;
+
+        if not {
+            self.write(" <> ")?;
+        } else {
+            self.write(" = ")?;
+        }
```
Comment (pimeys) [inline src/visitor/mysql.rs:365]: nitpick: spacing after this block.

File: src/visitor/postgres.rs [hunk 3]
```diff
@@ -262,6 +262,114 @@ impl<'a> Visitor<'a> for Postgres<'a> {
...
+    #[cfg(all(feature = "json", any(feature = "postgresql", feature = "mysql")))]
+    fn visit_json_array_contains(&mut self, left: Expression<'a>, right: Expression<'a>, not: bool) -> visitor::Result {
+        if not {
+            self.write("( NOT ")?;
+        }
+
+        self.visit_expression(left)?;
+        self.write(" @> ")?;
+        self.visit_expression(right)?;
+        if not {
```
Comment (pimeys) [inline src/visitor/postgres.rs:306]: nitpick: spacing before this block

File: src/visitor/postgres.rs [hunk 4]
```diff
@@ -262,6 +262,114 @@ impl<'a> Visitor<'a> for Postgres<'a> {
...
+    }
+
+    #[cfg(all(feature = "json", any(feature = "postgresql", feature = "mysql")))]
+    fn visit_json_type_equals(&mut self, left: Expression<'a>, json_type: JsonType) -> visitor::Result {
+        self.write("JSONB_TYPEOF")?;
+        self.write("(")?;
+        self.visit_expression(left)?;
+        self.write(")")?;
+        self.write(" = ")?;
+        match json_type {
```
Comment (pimeys) [inline src/visitor/postgres.rs:364]: nitpick: spacing before this block.

**Review by pimeys — APPROVED** [review]
Just small details in the comments, otherwise this looks great.

Please squash before merging!

File: src/ast/compare.rs [hunk 5]
```diff
@@ -44,6 +44,30 @@ pub enum Compare<'a> {
     /// Raw comparator, allows to use an operator `left <raw> right` as is,
     /// without visitor transformation in between.
     Raw(Box<Expression<'a>>, Cow<'a, str>, Box<Expression<'a>>),
+    #[cfg(all(feature = "json", any(feature = "postgresql", feature = "mysql")))]
```
Comment (dpetrick) [inline src/ast/compare.rs:47]: I feel like the JSON ops could have been encapsulated in another enum that in turn is included here. Blows up the compare AST too much for my taste.

**Review by dpetrick — APPROVED** [review]

File: src/ast/compare.rs (hunk 5 above)
Comment (pimeys) [reply src/ast/compare.rs:47]: yeah that looks much better now.

**Pull request merged**